### PR TITLE
MRDL->MRTK: Pointer focus locking

### PR DESCRIPTION
**Pull request by Railboy**
This PR adds a FocusLocked field to IPointingSource.

While a pointer's FocusLocked is true, focus enter / exit events are suspended and that pointer will continue to send manipulation to the last focused object.

This is needed for tools like bounding boxes, which require gaze focus for initial selection but not for subsequent manipulation.

PR passes all input tests.

### CODE REVIEW

**Comment by StephenHodgson** [comment]
@Railboy open this PR on my StephenHodgson:MRTK-uiRaycastHotfix branch

There's too many conflicts with #1350

**Comment by StephenHodgson** [comment]
Or at least hold off until we've fixed the issues we already have.

The master branch is currently very unstable bc of the ray stepping pr.

**Comment by Railboy** [comment]
What specific issues are we having that are caused by the ray stepping pr? Happy to help clean them up.

**Comment by Railboy** [comment]
In the future, let's not close pull requests if they're just being put on hold while other issues are being worked out. Other reviewers can still evaluate the code and raise potential issues in the meantime.

**Comment by StephenHodgson** [comment]
I'm by no means trying to subvert this PR, but instead asking you to open it on the beach I'm already working on, so we don't duplicate, overlap work.

The same reviewers should still be added.

If you don't feel like this is adequate them please remove the duplicated changes.

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs [hunk 1]
```diff
@@ -601,6 +609,9 @@ private void RaycastUnityUI(PointerData pointer, LayerMask[] prioritizedLayerMas
             RayStep rayStep = default(RayStep);
             int rayStepIndex = 0;
 
+            Debug.Assert(pointer.PointingSource.Rays != null);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: Nit: conflicts with changes in 1350

**Comment by StephenHodgson** [comment]
Oh I see, it's on a per pointer basis.

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs [hunk 2]
```diff
@@ -613,7 +624,7 @@ private void RaycastUnityUI(PointerData pointer, LayerMask[] prioritizedLayerMas
             }
 
             // Check if we need to overwrite the physics raycast info
-            if ((pointer.End.Object == null || overridePhysicsRaycast) && uiRaycastResult.module.eventCamera == UIRaycastCamera)
+            if ((pointer.End.Object == null || overridePhysicsRaycast) && (uiRaycastResult.module != null && uiRaycastResult.module.eventCamera == UIRaycastCamera))
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: This is also fixed in 1350

Comment (Railboy) [reply Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: Exactly, we want it to be undisruptive to other pointers.

>This is also fixed in 1350

Got it, I'll rebase once 1350 is merged.

Comment (StephenHodgson) [reply Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: `uiRaycastResult.IsValid`

File: Assets/HoloToolkit/Input/Scripts/Focus/RayStep.cs [hunk 3]
```diff
@@ -44,5 +44,94 @@ public void CopyRay (Ray ray, float rayLength)
...
+
+        /// <summary>
+        /// Returns a direction along an array of RaySteps by distance
+        /// </summary>
+        /// <param name="steps"></param>
+        /// <param name="distance"></param>
+        /// <returns></returns>
+        public static Vector3 GetDirectionByDistance(RayStep[] steps, float distance)
+        {
+            return GetStepByDistance(steps, distance).direction;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Input/Scripts/Focus/RayStep.cs]: should we also add the asserts here too?

**Comment by StephenHodgson** [comment]
Did you rerun the tests?

File: Assets/HoloToolkit/Input/Scripts/Focus/RayStep.cs (hunk 3 above)
Comment (Railboy) [reply Assets/HoloToolkit/Input/Scripts/Focus/RayStep.cs]: Good idea.

**Comment by Railboy** [comment]
All pass. Also tested in InputManagerTest, InteractableObjectExample & InteractiveButtonComponents scenes and everything appears normal.

**Pull request merged**